### PR TITLE
postgres fix: detect servers version and use the right query

### PR DESCRIPTION
**Pull request by ilyam8**
##### Summary
Fixes: #4910
Fixes: #4904

##### Component Name
python.d module `postgres`

##### Additional Information

add support for use specific query depending on server version

@anayrat 
@arenddeboer 

if you have free time please test, the issue should be fixed

@anayrat please look at all queries, I could break something :sweat_smile:

### CODE REVIEW

**Comment by anayrat** [comment]
Thanks @ilyam8! (unfortunately my server crashed, I am busy to reinstall it :( )
It is possible to have something like : 
```
if version >= 110000 
  (query for version 11)
else version >= 100000
  (query for version 10)
else
  ( default query)
```

It will be easier to handle different versions in the future than adding rules like: 
```
wal = 'wal' if version >= 100000 else 'xlog'
lsn = 'lsn' if version >= 100000 else 'location'
```
I am afraid queries will be hard to read after several version.
Thanks, and sorry to not help more :/

**Comment by ilyam8** [comment]
@anayrat done

**Comment by arenddeboer** [comment]
When testing I get this error:
```
2018-12-08 09:29:54: python.d DEBUG: postgres: socket: create() => [OK] (charts: 50)
2018-12-08 09:29:54: python.d DEBUG: postgres: socket: started, update frequency: 1
2018-12-08 09:29:54: python.d ERROR: postgres: socket: update() unhandled exception: operator does not exist: numeric / text
LINE 23: ...pg_walfile_name_offset (restart_lsn)).file_offset) / (s.val)
                                                               ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

**Comment by arenddeboer** [comment]
Adding a cast ```::INTEGER``` to both s.val instances seems to do the trick

**Comment by ilyam8** [comment]
@arenddeboer i am not familiar with postgres syntax please help.

`- (pg_walfile_name_offset (restart_lsn)).file_offset) / (s.val)`

=>
`- (pg_walfile_name_offset (restart_lsn)).file_offset) / (s.val)::INTEGER` 

? 
:sweat_smile:

**Comment by arenddeboer** [comment]
sorry, I have it like this:
```
(pg_walfile_name_offset (restart_lsn)).file_offset) / (s.val::INTEGER)
```

**Comment by ilyam8** [comment]
Thanks, let's wait for @anayrat

**Comment by ilyam8** [comment]
@anayrat 

what about 
https://github.com/netdata/netdata/pull/4944#issuecomment-445488188

is this the right way to do it?

**Comment by anayrat** [comment]
> @anayrat
> 
> what about
> [#4944 (comment)](https://github.com/netdata/netdata/pull/4944#issuecomment-445488188)
> 
> is this the right way to do it?

I am not sure, I don't understand why this query must be changed.

**Comment by ilyam8** [comment]
He got this error
https://github.com/netdata/netdata/pull/4944#issuecomment-445476167

> I will try to take time to test

ok, let's wait for your tests

File: collectors/python.d.plugin/postgres/postgres.chart.py [hunk 1]
```diff
@@ -296,13 +419,61 @@
     slot_type,
     replslot_wal_keep;
 """,
-    'IF_SUPERUSER': """
+    V11: """
+WITH wal_size AS (
```
Comment (anayrat) [inline collectors/python.d.plugin/postgres/postgres.chart.py]: You should replace : 
```
WITH wal_size AS (
          SELECT
            current_setting('wal_block_size')::INT * setting::INT AS val
          FROM pg_settings
          WHERE name = 'wal_segment_size'
          )
```

With : 
```
WITH wal_size AS (
          SELECT
             setting::int AS val
          FROM pg_settings
          WHERE name = 'wal_segment_size'
          )
```

Comment (ilyam8) [reply collectors/python.d.plugin/postgres/postgres.chart.py]: done

**Comment by anayrat** [comment]
I changed the logic to:
```
    elif name == WAL:
        if version < 100000:
            return QUERY_WAL[V96]
        return QUERY_WAL[DEFAULT]
    elif name == ARCHIVE:
        if version < 100000:
            return QUERY_ARCHIVE[V96]
        return QUERY_ARCHIVE[DEFAULT]
    elif name == STANDBY_DELTA:
        if version < 100000:
            return QUERY_STANDBY_DELTA[V96]
        return QUERY_STANDBY_DELTA[DEFAULT]
    elif name == REPSLOT_FILES:
        if version < 110000:
            return QUERY_REPSLOT_FILES[V10]
        return QUERY_REPSLOT_FILES[DEFAULT]
    elif name == DIFF_LSN:
        if version < 100000:
            return QUERY_DIFF_LSN[V96]
        return QUERY_DIFF_LSN[DEFAULT]
```

Now, default means latest postgres version and we handle previous version with `if` block.
Here is the diff : 
```diff
122a123
> V96 = 'V96'
136c137
<         pg_xlogfile_name(
---
>         pg_walfile_name(
139c140
<             ELSE pg_current_xlog_location()
---
>             ELSE pg_current_wal_lsn()
142c143
<           WHEN wal.name > pg_xlogfile_name(
---
>           WHEN wal.name > pg_walfile_name(
145c146
<               ELSE pg_current_xlog_location()
---
>               ELSE pg_current_wal_lsn()
149c150
<     FROM pg_catalog.pg_ls_dir('pg_xlog') AS wal(name)
---
>     FROM pg_catalog.pg_ls_dir('pg_wal') AS wal(name)
152c153
<         (pg_stat_file('pg_xlog/'||name)).modification,
---
>         (pg_stat_file('pg_wal/'||name)).modification,
155c156
<     V10: """
---
>     V96: """
163c164
<         pg_walfile_name(
---
>         pg_xlogfile_name(
166c167
<             ELSE pg_current_wal_lsn()
---
>             ELSE pg_current_xlog_location()
169c170
<           WHEN wal.name > pg_walfile_name(
---
>           WHEN wal.name > pg_xlogfile_name(
172c173
<               ELSE pg_current_wal_lsn()
---
>               ELSE pg_current_xlog_location()
176c177
<     FROM pg_catalog.pg_ls_dir('pg_wal') AS wal(name)
---
>     FROM pg_catalog.pg_ls_dir('pg_xlog') AS wal(name)
179c180
<         (pg_stat_file('pg_wal/'||name)).modification,
---
>         (pg_stat_file('pg_xlog/'||name)).modification,
191c192
<     pg_catalog.pg_ls_dir('pg_xlog/archive_status') AS archive_files (archive_file);
---
>     pg_catalog.pg_ls_dir('pg_wal/archive_status') AS archive_files (archive_file);
193c194
<     V10: """
---
>     V96: """
199c200,201
<     pg_catalog.pg_ls_dir('pg_wal/archive_status') AS archive_files (archive_file);
---
>     pg_catalog.pg_ls_dir('pg_xlog/archive_status') AS archive_files (archive_file);
> 
323,352d324
<     pg_xlog_location_diff(
<       CASE pg_is_in_recovery()
<         WHEN true THEN pg_last_xlog_receive_location()
<         ELSE pg_current_xlog_location()
<       END,
<     sent_location) AS sent_delta,
<     pg_xlog_location_diff(
<       CASE pg_is_in_recovery()
<         WHEN true THEN pg_last_xlog_receive_location()
<         ELSE pg_current_xlog_location()
<       END,
<     write_location) AS write_delta,
<     pg_xlog_location_diff(
<       CASE pg_is_in_recovery()
<         WHEN true THEN pg_last_xlog_receive_location()
<         ELSE pg_current_xlog_location()
<       END,
<     flush_location) AS flush_delta,
<     pg_xlog_location_diff(
<       CASE pg_is_in_recovery()
<         WHEN true THEN pg_last_xlog_receive_location()
<         ELSE pg_current_xlog_location()
<       END,
<     replay_location) AS replay_delta
< FROM pg_stat_replication
< WHERE application_name IS NOT NULL;
< """,
<     V10: """
< SELECT
<     application_name,
379a352,381
>     V96: """
> SELECT
>     application_name,
>     pg_xlog_location_diff(
>       CASE pg_is_in_recovery()
>         WHEN true THEN pg_last_xlog_receive_location()
>         ELSE pg_current_xlog_location()
>       END,
>     sent_location) AS sent_delta,
>     pg_xlog_location_diff(
>       CASE pg_is_in_recovery()
>         WHEN true THEN pg_last_xlog_receive_location()
>         ELSE pg_current_xlog_location()
>       END,
>     write_location) AS write_delta,
>     pg_xlog_location_diff(
>       CASE pg_is_in_recovery()
>         WHEN true THEN pg_last_xlog_receive_location()
>         ELSE pg_current_xlog_location()
>       END,
>     flush_location) AS flush_delta,
>     pg_xlog_location_diff(
>       CASE pg_is_in_recovery()
>         WHEN true THEN pg_last_xlog_receive_location()
>         ELSE pg_current_xlog_location()
>       END,
>     replay_location) AS replay_delta
> FROM pg_stat_replication
> WHERE application_name IS NOT NULL;
> """,
386c388
<     setting AS val
---
>     setting::int AS val
422c424
<     V11: """
---
>     V10: """
424,428c426,430
<           SELECT
<              setting::int AS val
<           FROM pg_settings
<           WHERE name = 'wal_segment_size'
<           )
---
>   SELECT
>     current_setting('wal_block_size')::INT * setting::INT AS val
>   FROM pg_settings
>   WHERE name = 'wal_segment_size'
>   )
493c495
<     pg_xlog_location_diff(
---
>     pg_wal_lsn_diff(
495,496c497,498
<         WHEN true THEN pg_last_xlog_receive_location()
<         ELSE pg_current_xlog_location()
---
>         WHEN true THEN pg_last_wal_receive_lsn()
>         ELSE pg_current_wal_lsn()
500c502
<     V10: """
---
>     V96: """
502c504
<     pg_wal_lsn_diff(
---
>     pg_xlog_location_diff(
504,505c506,507
<         WHEN true THEN pg_last_wal_receive_lsn()
<         ELSE pg_current_wal_lsn()
---
>         WHEN true THEN pg_last_xlog_receive_location()
>         ELSE pg_current_xlog_location()
538,539c540,541
<         if version >= 100000:
<             return QUERY_WAL[V10]
---
>         if version < 100000:
>             return QUERY_WAL[V96]
542,543c544,545
<         if version >= 100000:
<             return QUERY_ARCHIVE[V10]
---
>         if version < 100000:
>             return QUERY_ARCHIVE[V96]
546,547c548,549
<         if version >= 100000:
<             return QUERY_STANDBY_DELTA[V10]
---
>         if version < 100000:
>             return QUERY_STANDBY_DELTA[V96]
550,551c552,553
<         if version >= 110000:
<             return QUERY_REPSLOT_FILES[V11]
---
>         if version < 110000:
>             return QUERY_REPSLOT_FILES[V10]
554,555c556,557
<         if version >= 100000:
<             return QUERY_DIFF_LSN[V10]
---
>         if version < 100000:
>             return QUERY_DIFF_LSN[V96]
939c941
<                 self.queries[query_factory(REPSLOT_FILES)] = METRICS[REPSLOT_FILES]
---
>                 self.queries[query_factory(REPSLOT_FILES, self.server_version)] = METRICS[REPSLOT_FILES]
```

**Comment by anayrat** [comment]
I tested on theses versions:

- 9.3 (primary and secondary in replication)
- 9.4 (primary and secondary in replication)
- 9.5 (primary and secondary in replication)
- 9.6 (primary and secondary in replication)
- 10 (primary and secondary in replication)
- 11 (primary and secondary in replication)

**Comment by ilyam8** [comment]
@anayrat 

please use `git diff`

**Comment by anayrat** [comment]
```diff
diff --git a/collectors/python.d.plugin/postgres/postgres.chart.py b/collectors/python.d.plugin/postgres/postgres.chart.py
index 02444bd4..e920fdfa 100644
--- a/collectors/python.d.plugin/postgres/postgres.chart.py
+++ b/collectors/python.d.plugin/postgres/postgres.chart.py
@@ -120,6 +120,7 @@ METRICS = {
 
 NO_VERSION = 0
 DEFAULT = 'DEFAULT'
+V96 = 'V96'
 V10 = 'V10'
 V11 = 'V11'
 
@@ -133,26 +134,26 @@ SELECT
 FROM
     (SELECT
         wal.name,
-        pg_xlogfile_name(
+        pg_walfile_name(
           CASE pg_is_in_recovery()
             WHEN true THEN NULL
-            ELSE pg_current_xlog_location()
+            ELSE pg_current_wal_lsn()
           END ),
         CASE
-          WHEN wal.name > pg_xlogfile_name(
+          WHEN wal.name > pg_walfile_name(
             CASE pg_is_in_recovery()
               WHEN true THEN NULL
-              ELSE pg_current_xlog_location()
+              ELSE pg_current_wal_lsn()
             END ) THEN 'recycled'
           ELSE 'written'
         END AS type
-    FROM pg_catalog.pg_ls_dir('pg_xlog') AS wal(name)
+    FROM pg_catalog.pg_ls_dir('pg_wal') AS wal(name)
     WHERE name ~ '^[0-9A-F]{{24}}$'
     ORDER BY
-        (pg_stat_file('pg_xlog/'||name)).modification,
+        (pg_stat_file('pg_wal/'||name)).modification,
         wal.name DESC) sub;
 """,
-    V10: """
+    V96: """
 SELECT
     count(*) as total_wal,
     count(*) FILTER (WHERE type = 'recycled') AS recycled_wal,
@@ -160,23 +161,23 @@ SELECT
 FROM
     (SELECT
         wal.name,
-        pg_walfile_name(
+        pg_xlogfile_name(
           CASE pg_is_in_recovery()
             WHEN true THEN NULL
-            ELSE pg_current_wal_lsn()
+            ELSE pg_current_xlog_location()
           END ),
         CASE
-          WHEN wal.name > pg_walfile_name(
+          WHEN wal.name > pg_xlogfile_name(
             CASE pg_is_in_recovery()
               WHEN true THEN NULL
-              ELSE pg_current_wal_lsn()
+              ELSE pg_current_xlog_location()
             END ) THEN 'recycled'
           ELSE 'written'
         END AS type
-    FROM pg_catalog.pg_ls_dir('pg_wal') AS wal(name)
+    FROM pg_catalog.pg_ls_dir('pg_xlog') AS wal(name)
     WHERE name ~ '^[0-9A-F]{{24}}$'
     ORDER BY
-        (pg_stat_file('pg_wal/'||name)).modification,
+        (pg_stat_file('pg_xlog/'||name)).modification,
         wal.name DESC) sub;
 """,
 }
@@ -188,15 +189,16 @@ SELECT
     CAST(COALESCE(SUM(CAST(archive_file ~ $r$\.ready$$r$ as INT)),0) AS INT) AS ready_count,
     CAST(COALESCE(SUM(CAST(archive_file ~ $r$\.done$$r$ AS INT)),0) AS INT) AS done_count
 FROM
-    pg_catalog.pg_ls_dir('pg_xlog/archive_status') AS archive_files (archive_file);
+    pg_catalog.pg_ls_dir('pg_wal/archive_status') AS archive_files (archive_file);
 """,
-    V10: """
+    V96: """
 SELECT
     CAST(COUNT(*) AS INT) AS file_count,
     CAST(COALESCE(SUM(CAST(archive_file ~ $r$\.ready$$r$ as INT)),0) AS INT) AS ready_count,
     CAST(COALESCE(SUM(CAST(archive_file ~ $r$\.done$$r$ AS INT)),0) AS INT) AS done_count
 FROM
-    pg_catalog.pg_ls_dir('pg_wal/archive_status') AS archive_files (archive_file);
+    pg_catalog.pg_ls_dir('pg_xlog/archive_status') AS archive_files (archive_file);
+
 """,
 }
 
@@ -320,36 +322,6 @@ QUERY_STANDBY_DELTA = {
     DEFAULT: """
 SELECT
     application_name,
-    pg_xlog_location_diff(
-      CASE pg_is_in_recovery()
-        WHEN true THEN pg_last_xlog_receive_location()
-        ELSE pg_current_xlog_location()
-      END,
-    sent_location) AS sent_delta,
-    pg_xlog_location_diff(
-      CASE pg_is_in_recovery()
-        WHEN true THEN pg_last_xlog_receive_location()
-        ELSE pg_current_xlog_location()
-      END,
-    write_location) AS write_delta,
-    pg_xlog_location_diff(
-      CASE pg_is_in_recovery()
-        WHEN true THEN pg_last_xlog_receive_location()
-        ELSE pg_current_xlog_location()
-      END,
-    flush_location) AS flush_delta,
-    pg_xlog_location_diff(
-      CASE pg_is_in_recovery()
-        WHEN true THEN pg_last_xlog_receive_location()
-        ELSE pg_current_xlog_location()
-      END,
-    replay_location) AS replay_delta
-FROM pg_stat_replication
-WHERE application_name IS NOT NULL;
-""",
-    V10: """
-SELECT
-    application_name,
     pg_wal_lsn_diff(
       CASE pg_is_in_recovery()
         WHEN true THEN pg_last_wal_receive_lsn()
@@ -377,13 +349,43 @@ SELECT
 FROM pg_stat_replication
 WHERE application_name IS NOT NULL;
 """,
+    V96: """
+SELECT
+    application_name,
+    pg_xlog_location_diff(
+      CASE pg_is_in_recovery()
+        WHEN true THEN pg_last_xlog_receive_location()
+        ELSE pg_current_xlog_location()
+      END,
+    sent_location) AS sent_delta,
+    pg_xlog_location_diff(
+      CASE pg_is_in_recovery()
+        WHEN true THEN pg_last_xlog_receive_location()
+        ELSE pg_current_xlog_location()
+      END,
+    write_location) AS write_delta,
+    pg_xlog_location_diff(
+      CASE pg_is_in_recovery()
+        WHEN true THEN pg_last_xlog_receive_location()
+        ELSE pg_current_xlog_location()
+      END,
+    flush_location) AS flush_delta,
+    pg_xlog_location_diff(
+      CASE pg_is_in_recovery()
+        WHEN true THEN pg_last_xlog_receive_location()
+        ELSE pg_current_xlog_location()
+      END,
+    replay_location) AS replay_delta
+FROM pg_stat_replication
+WHERE application_name IS NOT NULL;
+""",
 }
 
 QUERY_REPSLOT_FILES = {
     DEFAULT: """
 WITH wal_size AS (
   SELECT
-    setting AS val
+    setting::int AS val
   FROM pg_settings
   WHERE name = 'wal_segment_size'
   )
@@ -419,13 +421,13 @@ GROUP BY
     slot_type,
     replslot_wal_keep;
 """,
-    V11: """
+    V10: """
 WITH wal_size AS (
-          SELECT
-             setting::int AS val
-          FROM pg_settings
-          WHERE name = 'wal_segment_size'
-          )
+  SELECT
+    current_setting('wal_block_size')::INT * setting::INT AS val
+  FROM pg_settings
+  WHERE name = 'wal_segment_size'
+  )
 SELECT
     slot_name,
     slot_type,
@@ -490,19 +492,19 @@ WHERE query NOT LIKE '%%pg_stat_activity%%';
 QUERY_DIFF_LSN = {
     DEFAULT: """
 SELECT
-    pg_xlog_location_diff(
+    pg_wal_lsn_diff(
       CASE pg_is_in_recovery()
-        WHEN true THEN pg_last_xlog_receive_location()
-        ELSE pg_current_xlog_location()
+        WHEN true THEN pg_last_wal_receive_lsn()
+        ELSE pg_current_wal_lsn()
       END,
     '0/0') as wal_writes ;
 """,
-    V10: """
+    V96: """
 SELECT
-    pg_wal_lsn_diff(
+    pg_xlog_location_diff(
       CASE pg_is_in_recovery()
-        WHEN true THEN pg_last_wal_receive_lsn()
-        ELSE pg_current_wal_lsn()
+        WHEN true THEN pg_last_xlog_receive_location()
+        ELSE pg_current_xlog_location()
       END,
     '0/0') as wal_writes ;
 """,
@@ -535,24 +537,24 @@ def query_factory(name, version=NO_VERSION):
     elif name == AUTOVACUUM:
         return QUERY_AUTOVACUUM[DEFAULT]
     elif name == WAL:
-        if version >= 100000:
-            return QUERY_WAL[V10]
+        if version < 100000:
+            return QUERY_WAL[V96]
         return QUERY_WAL[DEFAULT]
     elif name == ARCHIVE:
-        if version >= 100000:
-            return QUERY_ARCHIVE[V10]
+        if version < 100000:
+            return QUERY_ARCHIVE[V96]
         return QUERY_ARCHIVE[DEFAULT]
     elif name == STANDBY_DELTA:
-        if version >= 100000:
-            return QUERY_STANDBY_DELTA[V10]
+        if version < 100000:
+            return QUERY_STANDBY_DELTA[V96]
         return QUERY_STANDBY_DELTA[DEFAULT]
     elif name == REPSLOT_FILES:
-        if version >= 110000:
-            return QUERY_REPSLOT_FILES[V11]
+        if version < 110000:
+            return QUERY_REPSLOT_FILES[V10]
         return QUERY_REPSLOT_FILES[DEFAULT]
     elif name == DIFF_LSN:
-        if version >= 100000:
-            return QUERY_DIFF_LSN[V10]
+        if version < 100000:
+            return QUERY_DIFF_LSN[V96]
         return QUERY_DIFF_LSN[DEFAULT]
 
     raise ValueError('unknown query')
@@ -936,7 +938,7 @@ class Service(SimpleService):
                 self.queries[query_factory(WAL, self.server_version)] = METRICS[WAL]
 
             if self.server_version >= 100000:
-                self.queries[query_factory(REPSLOT_FILES)] = METRICS[REPSLOT_FILES]
+                self.queries[query_factory(REPSLOT_FILES, self.server_version)] = METRICS[REPSLOT_FILES]
 
         if self.server_version >= 90400:
             self.queries[query_factory(AUTOVACUUM)] = METRICS[AUTOVACUUM]

```

**Comment by ilyam8** [comment]
i've applied all your changes

**Comment by ilyam8** [comment]
@arenddeboer 
please test it again

**Comment by ilyam8** [comment]
The PR[ well tested](https://github.com/netdata/netdata/pull/4944#issuecomment-446360274), i think it is ready.

@anayrat anything else you want to add? i am about to merge it

**Comment by anayrat** [comment]
@ilyam8 you can merge and thanks for your work on code refactoring ;)

**Comment by ilyam8** [comment]
@anayrat thanks for you help again :+1:

**Comment by ilyam8** [comment]
@anayrat i think we need timeouts:
  * connection
  * query

I haven't tested anything, but i am realized we aren't using any timeouts for postgres.

What i found (on google) so far:

> connect_timeout
> Maximum wait for connection, in seconds (write as a decimal integer, e.g. 10). Zero, negative, or not specified means wait indefinitely. The minimum allowed timeout is 2 seconds, therefore a value of 1 is interpreted as 2. This timeout applies separately to each host name or IP address. For example, if you specify two hosts and connect_timeout is 5, each host will time out if no connection is made within 5 seconds, so the total time spent waiting for a connection might be up to 10 seconds.

For the query timeout i am not sure. Can you help me here?

**Comment by anayrat** [comment]
I did not test, maybe there is a default connection timeout in netdata?

+1 for query timeout, especially if we add support for custom queries. Look at `statement_timeout` https://www.postgresql.org/docs/current/runtime-config-client.html

**Comment by ilyam8** [comment]
> I did not test, maybe there is a default connection timeout in netdata?

`postgres` module is repsonsible for it. i've just tested it

> psycopg2.OperationalError: could not connect to server: Connection timed out
	Is the server running on host "127.0.0.1" and accepting
	TCP/IP connections on port 5432?

raised after 130 seconds.

I have no clue how to test query timeout :thinking: If you know how to test it, please do.

I will open a new PR for timers.

**Comment by anayrat** [comment]
https://stackoverflow.com/questions/19963954/set-transaction-query-timeout-in-psycopg2/20101377

**Pull request merged**